shell/server/mail.js

```diff
@@ -584,3 +607,221 @@ var getSmtpPool = function () {
+
+          resp.on('end', function() {
+            var content = {};
+            rpcResponse.content = content;
+
+            content.statusCode = statusInfo.code;
+            if ("content-encoding" in resp.headers) content.encoding = resp.headers["content-encoding"];
+            if ("content-language" in resp.headers) content.language = resp.headers["content-language"];
+            if ("content-type" in resp.headers) content.language = resp.headers["content-type"];
+            if ("content-disposition" in resp.headers) {
+              var disposition = resp.headers["content-disposition"];
+              var parts = disposition.split(';');
+              if (parts[0].toLowerCase().trim() === "attachment") {
+                parts.forEach(function (part) {
+                  var splitPart = part.split('=');
+                  if (splitPart[0].toLowerCase().trim() === "filename") {
+                    content.disposition = {download: splitPart[1].trim()};
+                  }
+                });
+              }
+            }
+
+            // TODO(someday): stream responses that need it
+            content.body = {}
+            content.body.bytes = Buffer.concat(buffers);
+
+            resolve(rpcResponse);
+          });
+          break;
+        case 'noContent':
+          var noContent = {};
+          rpcResponse.noContent = noContent;
+          noContent.setShouldResetForm = statusInfo.shouldResetForm;
+          resolve(rpcResponse);
+          break;
+        case 'redirect':
+          var redirect = {};
+          rpcResponse.redirect = redirect;
+          redirect.isPermanent = statusInfo.isPermanent;
+          redirect.switchToGet = statusInfo.switchToGet;
+          if ("location" in resp.headers) redirect.location = resp.headers["location"];
+          resolve(rpcResponse);
+          break;
+        case 'clientError':
+          var clientError = {};
+          rpcResponse.clientError = clientError;
+          clientError.statusCode = statusInfo.clientErrorCode;
+          clientError.descriptionHtml = statusInfo.descriptionHtml;
+          resolve(rpcResponse);
+          break;
+        case 'serverError':
+          var serverError = {};
+          rpcResponse.serverError = serverError;
+          clientError.descriptionHtml = statusInfo.descriptionHtml;
+          resolve(rpcResponse);
+          break;
+        default:
+          // ???
+          err = new Error("Invalid status code " + resp.statusCode + " received in response.");
+          err.nature = "localBug";
+          reject(err);
+          break;
+      }
+    });
+
+    req.on('error', function (e) {
+      e.nature = "networkFailure";
+      reject(e);
+    });
+
+    req.setTimeout(15000, function () {
+      req.abort();
+      err = new Error("Request timed out.");
+      err.nature = "localBug";
+      err.durability = "overloaded";
+      reject(err);
+    });
+
+    if (content) {
+      // TODO(someday): support streaming
```

Comment:
Outbound streaming is handled using a completely different set of method (`putStreaming` / `postStreaming`), so this comment probably doesn't belong here.